Add assignment support for Sequential

Stonesjtu opened the pull request:
Simply a copy from `ModuleList`.
I think the `Sequential` is just a *forward*able version of `ModuleList`, it makes sense to combine both into one class.

Stonesjtu committed:
Add assignment support for Sequential

apaszke commented on torch/nn/modules/container.py:
```diff
@@ -62,6 +60,9 @@ def __getitem__(self, idx):
             return next(it)
 
+    def __setitem__(self, idx, module):
+        return setattr(self, str(idx), module)
```
It's not guaranteed that `i`th module will be called `str(i)` - see the constructor that takes an OrderedDict

Stonesjtu replied:
Actually I dislike the `[]` access pattern because it's not consistency with normal `Module`. But here we can't access the `sub modules` in  container because the attribute name is a pure number `0`, `1` ..., how about naming the sub modules with pre-fix such as `sub_module_0`, `sub_module_1`.

apaszke replied:
I'm not sure if I follow. We allow users to specify their own names, and we should keep this contract, so no for renaming the modules. If you want to get `i`th module, you need to iterate over `_modules.values()` until you reach the `i`th value.

Stonesjtu replied:
Iterating the `values()` won't let me to modify the value at all. Instead I work it around by iterating the module name and set the corresponding value.

Stonesjtu committed:
Use i-th key to set the value

Simply manipulate the variable returned by iterator only refers to a new
module object, rather than modify the existing one.

Stonesjtu committed:
Add example for indexing of Sequential

apaszke commented on torch/nn/modules/container.py:
```diff
@@ -38,6 +39,12 @@ class Sequential(Module):
+        # Specify the module by index
+        model[1] = nn.Sigmoid()
+        # Specify the module by name if initialized with OrderedDict
+        model.relu1 = nn.LeakyReLU()
```
Can you please remove those examples? They won't render correctly, and I don't think they should be there. It's enough to show how to construct the module in the main docstring

apaszke commented on torch/nn/modules/container.py:
```diff
@@ -49,18 +56,23 @@ def __init__(self, *args):
+    def _get_item_by_idx(self, iterator, idx):
+        """Get the idx-th item of the iterator"""
+        size = len(self)
+        if not (-size <= idx < size):
```
no need to use parenthesis

apaszke commented on torch/nn/modules/container.py:
```diff
@@ -49,18 +56,23 @@ def __init__(self, *args):
-            for i in range(idx):
-                next(it)
-            return next(it)
+            return self._get_item_by_idx(self._modules.items(), idx)
```
This should have been `_modules.values()` like in the previous version of the code

Stonesjtu committed:
Address the review comments

apaszke reviewed (commented):
Looks good now, but I forgot to ask for tests for this feature. Can you please add them?

Stonesjtu committed:
Add test for setitem of Sequential

Stonesjtu commented:
How do you think of using pure `python list` rather than `OrderedDict` for named initialization in `Sequential`. e.g.

```python
items = list(args)
for item in items:
    if isinstance(item, tuple):
        ...
    else:
        ...
```

apaszke commented:
@pytorchbot test this please

apaszke commented:
No, let's not complicate the API. This looks good as is. Thanks!
